### ISSUE TITLE
Bug: [no-unnecessary-condition] checking for array index reports incorrectly

### ISSUE THREAD

**Issue by hornta** [typescript-eslint/typescript-eslint#8281]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.3&fileType=.tsx&code=MYewdgzgLgBAhgJwXAngLhtBBLMBzAbQF0YBeGYgbgFgAoO7AMxgApFkUCBGAZh4HYSpYTACuYACYBTRrikSAlDADedGDDoBfIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gdQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
const array: string[] = [];

if (array[1337] === undefined) {
  
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-condition": ["error"],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

I expect this rule to not warn because it's actually a needed check to do. I can't assume `array[1337]` exists and is set to a string.

### Actual Result

It's incorrectly reported as an error

**Comment by auvred** [comment]
Duplicate of #8272, #7505, #7118, #7086, #5225, #2883

Explanation from https://github.com/typescript-eslint/typescript-eslint/issues/8272#issuecomment-1900483208:

> See the [`noUncheckedIndexedAccess`](https://www.typescriptlang.org/tsconfig#noUncheckedIndexedAccess) compiler option.
> 
> The rule is powered by types. Withe the option turned off - the types say the variable `item` is of type `T`. So the rule reports. With the option turned on - the types would say the variable `item` is of type `T | undefined`. So the rule would not report.
